src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,168 @@
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Malte Grave - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <fcntl.h>
+#include <linux/can.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <vector>
+
+#include "CANLayer.h"
+#include "basecommfb.h"
+#include "comtypes.h"
+#include "datatype.h"
+#include "devlog.h"
+#include "forte_any.h"
+#include "forte_array.h"
+
+using namespace forte::com_infra;
+
+CCANComLayer::CCANComLayer(CComLayer *paUpperLayer, CBaseCommFB *paComFB) :
+    CComLayer(paUpperLayer, paComFB) {}
+
+EComResponse CCANComLayer::openConnection(char *paLayerParameter) {
+    EComResponse eRetVal = e_InitInvalidId;
+    CParameterParser parser(paLayerParameter, ';', eCANComParamterAmount);
+    if (eCANComParamterAmount != parser.parseParameters()) {
+        return eRetVal;
+    }
+
+    setupCANInternals(parser);
+
+    switch (mFb->getComServiceType()) {
+    case e_Server:
+        DEVLOG_ERROR("[CAN Layer] the layer only supports the PUB/SUB "
+                     "pattern. Please use a Publisher block to send data.");
+        eRetVal = e_InitTerminated;
+        break;
+    case e_Client:
+        DEVLOG_ERROR("[CAN Layer] the layer only supports the PUB/SUB "
+                     "pattern. Please use a Subscribe block to receive data.");
+        eRetVal = e_InitTerminated;
+        break;
+    case e_Publisher:
+        // is handled via sendData
+        break;
+    case e_Subscriber:
+        // is handled via recvData
+        break;
+    }
+
+    DEVLOG_DEBUG("[CAN Layer] Opening connection\n");
+
+    mCANSocket = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+    if (CFDSelectHandler::scmInvalidFileDescriptor == mCANSocket) {
+        DEVLOG_ERROR("[CAN Layer] Socket can't be created.\n");
+        return e_InitInvalidId;
+    }
+
+    std::strncpy(mIfr.ifr_name, mCANParam.interfaceName.c_str(), IFNAMSIZ - 1);
+    mIfr.ifr_name[IFNAMSIZ - 1] =
+        '\0'; // Ensure that null termination is added if input is too long.
+    auto err = ioctl(mCANSocket, SIOCGIFINDEX, &mIfr);
+    if (err != 0) {
+        DEVLOG_ERROR("[CAN Layer] Socket can be controlled.\n");
+        close(mCANSocket);
+        return eRetVal;
+    }
+
+    mAddr.can_family = AF_CAN;
+    mAddr.can_ifindex = mIfr.ifr_ifindex;
+
+    if (bind(mCANSocket, (struct sockaddr *)&mAddr, sizeof(mAddr)) < 0) {
+        DEVLOG_ERROR("[CAN Layer] Binding of the socket unsuccessful.\n");
+        close(mCANSocket);
+        return eRetVal;
+    }
+
+    this->getExtEvHandler<CFDSelectHandler>().addComCallback(mCANSocket, this);
+    DEVLOG_DEBUG("[CAN Layer] Callback handler added.\n");
+
+    mCANParam.CANId = mCANParam.CANId;
```

Comment:
here's something wrong

Reply:
Yeah, I see it now, that's pretty useless.